File: easy/66-Plus-One.py
from typing import List

class Solution:
    def plusOne(self, digits: List[int]) -> List[int]:
        """Add 1 to the input number provided as a list"""
        digits[-1] += 1 # Adds 1 to the last digit

        # Loop back through each digit but the first
        for i in range(len(digits) - 1, 0, -1):
            if digits[i] > 9: # If there is a carry
                digits[i] -= 10 # Remove the carry from this digit
                digits[i-1] += 1 # Move the carry to the next digit
        
        if digits[0] > 9: # If there is a carry for the first digit
            digits[0] -= 10 # Remove the carry from this digit
            digits = [1] + digits # Add [1] to the front of the list

        return digits

File: easy/66-Plus-One.hs
-- Recursive function
recur :: [Int] -> [Int] -> [Int]
-- Singleton list case
recur (x:[]) [] = if x > 9 then [1,x-10] -- If digit is above 9
                  else [x] -- If digit is below 10
-- Base case for non-singleton lists
recur (x:[]) ys = if x > 9 then [1,x-10] ++ ys -- If digit is above 9
                  else x:ys -- If digit is below 10
-- Recursive case for non-singleton lists
recur xs ys = if (head (reverse xs)) > 9 then recur ((init (init xs)) ++ [((reverse xs) !! 1) + 1]) ([0] ++ ys) -- If digit is above 9
              else recur (init xs) ([(head (reverse xs))] ++ ys) -- If digit is below 10

-- Add 1 to the input number provided as a list
plusOne :: [Int] -> [Int]
plusOne xs = recur ((init xs) ++ [(head (reverse xs)) + 1]) [] -- Adding 1 to the last element in the list